### PR TITLE
Device Infiniband

### PR DESCRIPTION
**Pull request by tomponline**
Migration of infiniband device into the new devices interface.

- [x] Infiniband Physical
- [x] Infiniband SR-IOV
- [x] Lantea testing & scripted testing

Includes https://github.com/lxc/lxd/pull/6043

Part of #5819

### CODE REVIEW

**Comment by tomponline** [comment]
@stgraber this is ready for review now.

File: lxd/container_lxc.go [hunk 1]
```diff
@@ -1908,6 +1924,42 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
...
+		}
+	}
+
+	return nil
+}
+
+// deviceAddCgroupRules live adds cgroup rules to a container.
+func (c *containerLXC) deviceAddCgroupRules(configCopy map[string]string, cgroups []device.RunConfigItem) error {
+	for _, rule := range cgroups {
+		if c.isCurrentlyPrivileged() && !c.state.OS.RunningInUserNS && c.state.OS.CGroupDevicesController {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Just a nit (not blocker), but this should check that the key is indeed a devices cgroup key and only apply the restriction if that's the case.

We may soon get a gpu cgroup which would also be configured through the devices and would not have the same restriction as the devices croup.

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -1908,6 +1924,42 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
...
+			// Add the new device cgroup rule.
+			if err := c.CGroupSet(rule.Key, rule.Value); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Syntax nit, we prefer two lines.

File: lxd/device/device_instance_id.go [hunk 3]
```diff
@@ -15,4 +16,5 @@ type InstanceIdentifier interface {
 	LogPath() string
 	ExpandedConfig() map[string]string
 	ExpandedDevices() config.Devices
+	CurrentIdmap() (*idmap.IdmapSet, error)
```
Comment (stgraber) [inline lxd/device/device_instance_id.go]: I suspect we'll need to abstract that further later as that's a rather container-specific concept.

File: test/suites/container_devices_infiniband_physical.sh [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+# cd MLNX_OFED_LINUX-4.6-1.0.1.1-ubuntu18.04-x86_64/
+# sudo ./mlnxofedinstall  --force
+# sudo mlxconfig --yes -d /dev/mst/mt4099_pciconf0 set LINK_TYPE_P2=2
+# echo "options mlx4_core num_vfs=4 probe_vf=4" | sudo tee /etc/modprobe.d/mellanox.conf
+# reboot
+test_container_devices_ifb_physical() {
+  ensure_import_testimage
+  ensure_has_localhost_remote "${LXD_ADDR}"
+
+  parent=${LXD_IFB_PHYSICAL_PARENT:-""}
```
Comment (stgraber) [inline test/suites/container_devices_infiniband_physical.sh]: - standard abbreviation of infiniband is `IB`
 - New variable should be documented in test/README.md

**Comment by stgraber** [comment]
Looking good, just a few nits/comments.

File: lxd/container_lxc.go (hunk 1 above)
Comment (tomponline) [reply lxd/container_lxc.go]: @stgraber thats fine, I'll add a prefix match on "devices.". What is the best way to test that these are being set correctly?

Comment (tomponline) [reply lxd/container_lxc.go]: @stgraber I've used this approach:

1. Run the container as privileged: ```lxc config set c1 security.privileged true```

Then start the container and check the IB devices; major numbers are allowed in the container's cgroup devices.list:

```
root@lantea:/sys/fs/cgroup/devices/lxc.payload/c1# cat devices.list
b *:* m
c *:* m
c 136:* rwm
c 1:3 rwm
c 1:5 rwm
c 1:7 rwm
c 1:8 rwm
c 1:9 rwm
c 5:0 rwm
c 5:1 rwm
c 5:2 rwm
c 10:229 rwm
c 10:200 rwm
c 231:193 rwm
c 231:66 rwm
c 231:2 rwm
root@lantea:/sys/fs/cgroup/devices/lxc.payload/c1# cat /proc/devices 
Character devices:
  1 mem
  4 /dev/vc/0
  4 tty
  4 ttyS
  5 /dev/tty
  5 /dev/console
  5 /dev/ptmx
  5 ttyprintk
  7 vcs
 10 misc
 13 input
 21 sg
 29 fb
 89 i2c
108 ppp
128 ptm
136 pts
180 usb
189 usb_device
204 ttyMAX
226 drm
231 infiniband_mad
231 infiniband_verbs
```

File: lxd/device/device_instance_id.go (hunk 3 above)
Comment (tomponline) [reply lxd/device/device_instance_id.go]: @stgraber no problem, I've removed that from the interface and added shifting support to the `c.deviceStart` function by adding a new `c.deviceShiftMounts` function, which will then do the same thing as what `device.UnixCreateDevice` does when passed a non-nil idmap set.

**Comment by tomponline** [comment]
@stgraber this is ready for review again.

File: lxd/container_lxc.go [hunk 5]
```diff
@@ -1889,25 +1889,112 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
...
+	idmapSet, err := c.CurrentIdmap()
+	if err != nil {
+		return fmt.Errorf("Failed to get idmap for device: %s", err)
+	}
+
+	// If there is an idmap being applied and LXD not running in a user namespace then shift the
+	// device files before they are mounted.
+	if idmapSet != nil && !c.state.OS.RunningInUserNS {
+		for _, mount := range mounts {
+			if err := idmapSet.ShiftFile(mount.DevPath); err != nil {
```
Comment (stgraber) [inline lxd/container_lxc.go]: here

File: lxd/device/device_utils_disk.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+			}
+		}
+
+		if recursive {
+			flags |= unix.MS_REC
+		}
+	}
+
+	// Mount the filesystem
+	if err = unix.Mount(srcPath, dstPath, fstype, uintptr(flags), ""); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_disk.go]: here

File: lxd/device/device_utils_disk.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		return fmt.Errorf("Unable to mount %s at %s: %s", srcPath, dstPath, err)
+	}
+
+	// Remount bind mounts in readonly mode if requested
+	if readonly == true && flags&unix.MS_BIND == unix.MS_BIND {
+		flags = unix.MS_RDONLY | unix.MS_BIND | unix.MS_REMOUNT
+		if err = unix.Mount("", dstPath, fstype, uintptr(flags), ""); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_disk.go]: here

File: lxd/device/device_utils_disk.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+			return fmt.Errorf("Unable to mount %s in readonly mode: %s", dstPath, err)
+		}
+	}
+
+	flags = unix.MS_REC | unix.MS_SLAVE
+	if err = unix.Mount("", dstPath, "", uintptr(flags), ""); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_disk.go]: here

File: lxd/device/device_utils_unix.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+	devPath := filepath.Join(devicesPath, devName)
+
+	// Create the new entry
+	if !s.OS.RunningInUserNS {
+		encodedDeviceNumber := (minor & 0xff) | (major << 8) | ((minor & ^0xff) << 12)
+		if err := unix.Mknod(devPath, uint32(mode), encodedDeviceNumber); err != nil {
+			return nil, fmt.Errorf("Failed to create device %s for %s: %s", devPath, m["path"], err)
+		}
+
+		if err := os.Chown(devPath, uid, gid); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_unix.go]: here

File: lxd/device/device_utils_unix.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+		destPath = m["source"]
+	}
+	relativeDestPath := strings.TrimPrefix(destPath, "/")
+	devName := fmt.Sprintf("%s.%s", strings.Replace(prefix, "/", "-", -1), strings.Replace(relativeDestPath, "/", "-", -1))
+	devPath := filepath.Join(devicesPath, devName)
+
+	// Create the new entry
+	if !s.OS.RunningInUserNS {
+		encodedDeviceNumber := (minor & 0xff) | (major << 8) | ((minor & ^0xff) << 12)
+		if err := unix.Mknod(devPath, uint32(mode), encodedDeviceNumber); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_unix.go]: here

File: lxd/device/device_utils_unix.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+			return nil, fmt.Errorf("Failed to create device %s for %s: %s", devPath, m["path"], err)
+		}
+
+		if err := os.Chown(devPath, uid, gid); err != nil {
+			return nil, fmt.Errorf("Failed to chown device %s: %s", devPath, err)
+		}
+
+		// Needed as mknod respects the umask
+		if err := os.Chmod(devPath, mode); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_unix.go]: here

File: lxd/device/device_utils_unix.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+			return nil, fmt.Errorf("Failed to chmod device %s: %s", devPath, err)
+		}
+
+		if idmapSet != nil {
+			if err := idmapSet.ShiftFile(devPath); err != nil {
```
Comment (stgraber) [inline lxd/device/device_utils_unix.go]: here

File: lxd/container_lxc.go [hunk 13]
```diff
@@ -1889,25 +1889,112 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
...
+	}
+
+	return nil
+}
+
+// deviceAddCgroupRules live adds cgroup rules to a container.
+func (c *containerLXC) deviceAddCgroupRules(configCopy map[string]string, cgroups []device.RunConfigItem) error {
+	for _, rule := range cgroups {
+		// If container is running privileged and cgroup rule is for devices cgroup, apply the requested rule.
+		if c.isCurrentlyPrivileged() && !c.state.OS.RunningInUserNS && c.state.OS.CGroupDevicesController && strings.HasPrefix(rule.Key, "devices.") {
```
Comment (stgraber) [inline lxd/container_lxc.go]: What I meant with my comment is that we may soon find ourselves with calls to deviceAddCgroupRules which will not exclusively be made of `devices.` keys.

So this function should process them all and only if the prefix is devices. then skip when inside a userns or an unprivileged container.

Comment (tomponline) [reply lxd/container_lxc.go]: @stgraber I've made the change as we discussed.

**Pull request merged**